File: solutions/ruby/collatz-conjecture/3/collatz_conjecture.rb
class CollatzConjecture

  def self.steps number
    raise ArgumentError if number <= 0

    result = 0

    while number != 1
      # divide by two if even, otherwise multiply by three and add one
      number = number.even? ? number / 2 : number * 3 + 1
      result += 1
    end

    result
  end

end

File: solutions/elixir/collatz-conjecture/3/lib/collatz_conjecture.ex
defmodule CollatzConjecture do

  def calc(number) when is_integer(number) and number >= 1,
    do: steps(number, 0)

  defp steps(1, steps), do: steps

  defp steps(number, steps),
    do:
      steps(
        if(rem(number, 2) == 0,
          do: div(number, 2),
          else: 3 * number + 1),
        steps + 1)

end
